File: src/ui/toolbar.py
"""
顶部工具栏组件：串口和 UDP 连接控制。
"""

import logging

from PySide6.QtWidgets import (
    QWidget, QHBoxLayout, QLabel, QPushButton, QSpinBox,
    QComboBox, QStyledItemDelegate, QAbstractItemView,
)
from PySide6.QtCore import Qt, Signal, QSize
import serial.tools.list_ports

from src.ui.styles import (
    STYLE_LABEL, STYLE_COMBO, STYLE_SPINBOX,
    STYLE_BTN_IDLE, STYLE_BTN_ACTIVE, TOP_BAR_STYLE,
)

logger = logging.getLogger(__name__)


class _CompactItemDelegate(QStyledItemDelegate):
    """Item delegate that enforces a fixed row height in combo box dropdowns."""

    def __init__(self, height=20, parent=None):
        super().__init__(parent)
        self._height = height

    def sizeHint(self, option, index):
        size = super().sizeHint(option, index)
        size.setHeight(self._height)
        return size


class _SeamlessComboBox(QComboBox):
    """QComboBox whose popup tucks under the rounded bottom corners."""

    _radius = 4

    def showPopup(self):
        super().showPopup()
        popup = self.view().window()
        geo = popup.geometry()
        geo.setTop(geo.top() - self._radius)
        popup.setGeometry(geo)
        if popup.layout():
            popup.layout().setContentsMargins(0, self._radius, 0, 0)


class ToolBar(QWidget):
    """顶部工具栏，提供串口和 UDP 连接控制。"""

    serial_start_requested = Signal(str, int, str, int, int, int)
    serial_stop_requested = Signal()
    udp_start_requested = Signal(str, int)
    udp_stop_requested = Signal()

    def __init__(self, config_loader, data_receiver, parent=None):
        super().__init__(parent)
        self.config_loader = config_loader
        self.data_receiver = data_receiver

        self.setFixedHeight(32)
        self.setStyleSheet(TOP_BAR_STYLE)

        layout = QHBoxLayout(self)
        layout.setContentsMargins(10, 0, 10, 0)
        layout.setSpacing(8)

        ctrl_h = 24

        # ── Serial section ──
        serial_label = QLabel("串口 ")
        serial_label.setStyleSheet(STYLE_LABEL)
        layout.addWidget(serial_label)

        self.serial_combo = _SeamlessComboBox()
        self.serial_combo.setFixedHeight(ctrl_h)
        self.serial_combo.setStyleSheet(STYLE_COMBO)
        self.serial_combo.setItemDelegate(
            _CompactItemDelegate(20, self.serial_combo),
        )

        popup_window = self.serial_combo.view().window()
        popup_window.setWindowFlags(
            Qt.Popup | Qt.FramelessWindowHint | Qt.NoDropShadowWindowHint,
        )
        popup_window.setAttribute(Qt.WA_TranslucentBackground)
        self.serial_combo.view().setVerticalScrollMode(
            QAbstractItemView.ScrollMode.ScrollPerPixel,
        )
        self.serial_combo.view().setContentsMargins(0, 0, 0, 0)
        self.serial_combo.view().setSpacing(0)

        self._refresh_serial_ports()
        layout.addWidget(self.serial_combo)

        self.serial_refresh_btn = QPushButton("↻")
        self.serial_refresh_btn.setFixedSize(ctrl_h, ctrl_h)
        self.serial_refresh_btn.setStyleSheet(
            STYLE_BTN_IDLE
            + "QPushButton { font-size: 14px; font-weight: bold; padding: 0px; }",
        )
        self.serial_refresh_btn.setToolTip("刷新串口列表")
        self.serial_refresh_btn.clicked.connect(self._refresh_serial_ports)
        layout.addWidget(self.serial_refresh_btn)

        self.serial_toggle_btn = QPushButton("打开串口")
        self.serial_toggle_btn.setFixedHeight(ctrl_h)
        self.serial_toggle_btn.setStyleSheet(STYLE_BTN_IDLE)
        self.serial_toggle_btn.clicked.connect(self._toggle_serial)
        layout.addWidget(self.serial_toggle_btn)

        layout.addSpacing(16)

        # ── UDP section ──
        udp_port_label = QLabel("UDP端口 ")
        udp_port_label.setStyleSheet(STYLE_LABEL)
        layout.addWidget(udp_port_label)

        udp_config = self.config_loader.get_udp_config()
        self.udp_port_spin = QSpinBox()
        self.udp_port_spin.setRange(1, 65535)
        self.udp_port_spin.setValue(udp_config.get("port", 8888))
        self.udp_port_spin.setFixedHeight(ctrl_h)
        self.udp_port_spin.setFixedWidth(72)
        self.udp_port_spin.setStyleSheet(STYLE_SPINBOX)
        layout.addWidget(self.udp_port_spin)

        self.udp_toggle_btn = QPushButton("接收UDP")
        self.udp_toggle_btn.setFixedHeight(ctrl_h)
        self.udp_toggle_btn.setStyleSheet(STYLE_BTN_IDLE)
        self.udp_toggle_btn.clicked.connect(self._toggle_udp)
        layout.addWidget(self.udp_toggle_btn)

        layout.addStretch()

        # 串口异步断开时更新 UI
        self.data_receiver.serial_stopped.connect(self._on_serial_stopped)

    # ── Serial ───────────────────────────────────────────────────────

    def _refresh_serial_ports(self):
        current = self.serial_combo.currentText()
        self.serial_combo.clear()
        ports = serial.tools.list_ports.comports()
        serial_config = self.config_loader.get_serial_config()
        cfg_port = serial_config.get("port", "")
        select_idx = 0
        for i, info in enumerate(sorted(ports, key=lambda p: p.device)):
            label = (
                f"{info.device}  {info.description}"
                if info.description and info.description != "n/a"
                else info.device
            )
            self.serial_combo.addItem(label, userData=info.device)
            if info.device == current or (
                not current and info.device == cfg_port
            ):
                select_idx = i
        if self.serial_combo.count() > 0:
            self.serial_combo.setCurrentIndex(select_idx)

    def _toggle_serial(self):
        if self.data_receiver.is_serial_running:
            self.data_receiver.stop_serial()
            self._set_serial_ui_idle()
        else:
            if self.serial_combo.count() == 0:
                return
            port = self.serial_combo.currentData()
            serial_cfg = self.config_loader.get_serial_config()
            self.data_receiver.start_serial(
                port=port,
                baudrate=serial_cfg.get("baudrate", 115200),
                protocol=serial_cfg.get("protocol", "csv"),
                timeout=serial_cfg.get("timeout", 1),
                acc_fsr=serial_cfg.get("acc_fsr", 4),
                gyro_fsr=serial_cfg.get("gyro_fsr", 2000),
            )
            self.serial_toggle_btn.setText("关闭串口")
            self.serial_toggle_btn.setStyleSheet(STYLE_BTN_ACTIVE)
            self.serial_combo.setEnabled(False)
            self.serial_refresh_btn.setEnabled(False)

    def _set_serial_ui_idle(self):
        self.serial_toggle_btn.setText("打开串口")
        self.serial_toggle_btn.setStyleSheet(STYLE_BTN_IDLE)
        self.serial_combo.setEnabled(True)
        self.serial_refresh_btn.setEnabled(True)
        self.serial_stop_requested.emit()

    def _on_serial_stopped(self):
        """Handle asynchronous serial disconnect."""
        self._set_serial_ui_idle()

    # ── UDP ──────────────────────────────────────────────────────────

    def _toggle_udp(self):
        if self.data_receiver.is_udp_running:
            self.data_receiver.stop_udp()
            self.udp_toggle_btn.setText("接收UDP")
            self.udp_toggle_btn.setStyleSheet(STYLE_BTN_IDLE)
            self.udp_port_spin.setEnabled(True)
        else:
            udp_cfg = self.config_loader.get_udp_config()
            ip = udp_cfg.get("ip", "127.0.0.1")
            port = self.udp_port_spin.value()
            self.data_receiver.start_udp(ip, port)
            self.udp_toggle_btn.setText("停止接收")
            self.udp_toggle_btn.setStyleSheet(STYLE_BTN_ACTIVE)
            self.udp_port_spin.setEnabled(False)

File: src/ui/ToolBar.cpp
#include "ToolBar.h"
#include "Styles.h"
#include "../io/DataReceiver.h"
#include "../config/ConfigLoader.h"

#include <QLabel>
#include <QStyledItemDelegate>
#include <QAbstractItemView>
#include <QListView>
#include <QTimer>
#include <QHBoxLayout>
#include <QFontMetrics>
#include <QStyle>
#include <QStyleOptionComboBox>
#include <algorithm>
#ifdef HAVE_QT_SERIAL_PORT
#  include <QSerialPortInfo>
#endif

// ============================================================
// ToolBar.cpp — 顶部工具栏实现
// ============================================================

// ── SeamlessComboBox ─────────────────────────────────────────

SeamlessComboBox::SeamlessComboBox(QWidget* parent) : QComboBox(parent) {}

// 弹出列表齐平放在选框正下方，无重叠。
void SeamlessComboBox::showPopup() {
    QComboBox::showPopup();
    QWidget* popup = view()->window();
    if (!popup)
        return;

    QRect geo = popup->geometry();
    geo.moveTopLeft(mapToGlobal(QPoint(0, height())));
    geo.setWidth(std::max(geo.width(), width()));
    popup->setGeometry(geo);
}

// ── CompactItemDelegate ───────────────────────────────────────

class CompactItemDelegate : public QStyledItemDelegate {
public:
    explicit CompactItemDelegate(int height, QObject* parent = nullptr)
        : QStyledItemDelegate(parent), m_height(height) {}

    // 强制每行高度为固定值
    QSize sizeHint(const QStyleOptionViewItem& option, const QModelIndex& index) const override {
        QSize s = QStyledItemDelegate::sizeHint(option, index);
        s.setHeight(m_height);
        return s;
    }
private:
    int m_height;
};

// ── ToolBar ───────────────────────────────────────────────────

ToolBar::ToolBar(QWidget* parent) : QWidget(parent) {
    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(8);

    const int ctrlH = 24;

    // ── 串口区段 ──
    auto* serialLabel = new QLabel("串口 ");
    serialLabel->setStyleSheet(Styles::STYLE_LABEL());
    layout->addWidget(serialLabel);

    m_serialCombo = new SeamlessComboBox(this);
    m_serialCombo->setFixedHeight(ctrlH);
    m_serialCombo->setStyleSheet(Styles::STYLE_COMBO());
    m_serialCombo->setItemDelegate(new CompactItemDelegate(20, m_serialCombo));
    m_serialCombo->setSizeAdjustPolicy(QComboBox::AdjustToContents);
    m_serialCombo->view()->setVerticalScrollMode(QAbstractItemView::ScrollPerPixel);
    m_serialCombo->view()->setContentsMargins(0, 0, 0, 0);
    // setSpacing 是 QListView 的成员，需要向下转型
    if (auto* lv = qobject_cast<QListView*>(m_serialCombo->view()))
        lv->setSpacing(0);
    layout->addWidget(m_serialCombo);

    m_serialToggle = new QPushButton("打开串口", this);
    m_serialToggle->setFixedHeight(ctrlH);
    m_serialToggle->setStyleSheet(Styles::STYLE_BTN_IDLE());
    connect(m_serialToggle, &QPushButton::clicked, this, &ToolBar::toggleSerial);
    layout->addWidget(m_serialToggle);

    layout->addSpacing(16);

    // ── UDP 区段 ──
    auto* udpLabel = new QLabel("UDP端口 ");
    udpLabel->setStyleSheet(Styles::STYLE_LABEL());
    layout->addWidget(udpLabel);

    auto& cfg = ConfigLoader::instance();
    UdpConfig udpCfg = cfg.getUdpConfig();

    m_udpPortSpin = new QSpinBox(this);
    m_udpPortSpin->setRange(1, 65535);
    m_udpPortSpin->setValue(udpCfg.port);
    m_udpPortSpin->setFixedHeight(ctrlH);
    m_udpPortSpin->setFixedWidth(72);
    m_udpPortSpin->setStyleSheet(Styles::STYLE_SPINBOX());
    layout->addWidget(m_udpPortSpin);

    m_udpToggle = new QPushButton("接收UDP", this);
    m_udpToggle->setFixedHeight(ctrlH);
    m_udpToggle->setStyleSheet(Styles::STYLE_BTN_IDLE());
    connect(m_udpToggle, &QPushButton::clicked, this, &ToolBar::toggleUdp);
    layout->addWidget(m_udpToggle);

#ifdef HAVE_QT_SERIAL_PORT
    m_serialPollTimer = new QTimer(this);
    m_serialPollTimer->setInterval(1000);
    connect(m_serialPollTimer, &QTimer::timeout, this, &ToolBar::onSerialPollTimer);
#endif

    // 初始化串口下拉列表（有模块时随后启动轮询）
    updateSerialPortList();
    QTimer::singleShot(0, this, &ToolBar::lockSerialComboSize);
#ifdef HAVE_QT_SERIAL_PORT
    m_serialPollTimer->start();
#endif
}

// 绑定数据接收器，连接异步断开信号
void ToolBar::bindDataReceiver(DataReceiver* recv) {
    m_recv = recv;
    if (m_recv) {
        connect(m_recv, &DataReceiver::serialStopped, this, &ToolBar::onSerialStopped);
    }
}

void ToolBar::onSerialPollTimer() {
#ifdef HAVE_QT_SERIAL_PORT
    if (m_recv && m_recv->isSerialRunning())
        return;
    updateSerialPortList();
#endif
}

void ToolBar::lockSerialComboSize() {
    if (!m_serialCombo || m_serialComboSizeLocked)
        return;

    const int lockedHeight = std::max(m_serialCombo->height(), m_serialCombo->sizeHint().height());
    m_serialCombo->setFixedSize(serialComboLockedWidth(), lockedHeight);
    m_serialComboSizeLocked = true;
}

int ToolBar::serialComboLockedWidth() const {
    if (!m_serialCombo)
        return 120;

    QString widestText = m_serialCombo->currentText();
    QFontMetrics metrics(m_serialCombo->font());
    int widestTextPx = metrics.horizontalAdvance(widestText);

    for (int i = 0; i < m_serialCombo->count(); ++i) {
        const QString itemText = m_serialCombo->itemText(i);
        const int itemWidth = metrics.horizontalAdvance(itemText);
        if (itemWidth > widestTextPx) {
            widestTextPx = itemWidth;
            widestText = itemText;
        }
    }

    QStyleOptionComboBox opt;
    opt.initFrom(m_serialCombo);
    opt.currentText = widestText;
    opt.editable = m_serialCombo->isEditable();

    const QSize contentsSize(widestTextPx + 28, m_serialCombo->sizeHint().height());
    const int width = m_serialCombo->style()->sizeFromContents(
        QStyle::CT_ComboBox, &opt, contentsSize, m_serialCombo).width();
    return std::max(width, m_serialCombo->minimumSizeHint().width());
}

// 枚举串口并刷新下拉列表；若相对上次轮询出现新端口且非首轮，则自动选中最晚枚举到的新端口。
void ToolBar::updateSerialPortList() {
    auto& cfg = ConfigLoader::instance();
    SerialConfig sCfg = cfg.getSerialConfig();
    const QString cfgPort = sCfg.port;

#ifdef HAVE_QT_SERIAL_PORT
    QList<QSerialPortInfo> portsRaw = QSerialPortInfo::availablePorts();
    QSet<QString> nowNames;
    for (const auto& info : portsRaw)
        nowNames.insert(info.portName());

    // 端口集合未变则不碰下拉框，避免每秒 clear/add 引发整窗重绘与 3D 卡顿。
    if (nowNames == m_serialPortsKnown)
        return;

    const QString current = m_serialCombo->currentData().toString();
    m_serialCombo->clear();

    const QSet<QString> prevKnown = m_serialPortsKnown;

    QString preferredNewPort;
    if (!prevKnown.isEmpty()) {
        // 按 availablePorts() 顺序：同批多次出现的新口时，后以系统枚举顺序为准（通常新插入靠后）
        for (const auto& info : portsRaw) {
            if (!prevKnown.contains(info.portName()))
                preferredNewPort = info.portName();
        }
    }

    QList<QSerialPortInfo> ports = portsRaw;
    std::sort(ports.begin(), ports.end(), [](const QSerialPortInfo& a, const QSerialPortInfo& b){
        return a.portName() < b.portName();
    });

    for (int i = 0; i < ports.size(); ++i) {
        const auto& info = ports[i];
        QString label = (!info.description().isEmpty() && info.description() != "n/a")
            ? QString("%1  %2").arg(info.portName(), info.description())
            : info.portName();
        m_serialCombo->addItem(label, info.portName());
    }

    int targetIdx = 0;
    bool picked = false;
    if (!preferredNewPort.isEmpty()) {
        for (int i = 0; i < m_serialCombo->count(); ++i) {
            if (m_serialCombo->itemData(i).toString() == preferredNewPort) {
                targetIdx = i;
                picked = true;
                break;
            }
        }
    }
    if (!picked) {
        for (int i = 0; i < m_serialCombo->count(); ++i) {
            const QString name = m_serialCombo->itemData(i).toString();
            if (name == current || (current.isEmpty() && name == cfgPort)) {
                targetIdx = i;
                picked = true;
                break;
            }
        }
    }
    if (!picked && m_serialCombo->count() > 0)
        targetIdx = 0;

    if (m_serialCombo->count() > 0)
        m_serialCombo->setCurrentIndex(targetIdx);

    m_serialPortsKnown = nowNames;
#else
    m_serialCombo->clear();
    m_serialCombo->addItem("(SerialPort 模块未安装)", QString());
    m_serialCombo->setEnabled(false);
    m_serialToggle->setEnabled(false);
    Q_UNUSED(cfgPort);
#endif

    if (!m_serialComboSizeLocked)
        lockSerialComboSize();
}

// 切换串口连接状态
void ToolBar::toggleSerial() {
    if (!m_recv) return;

    if (m_recv->isSerialRunning()) {
        // 已连接 → 断开
        m_recv->stopSerial();
        setSerialUiIdle();
    } else {
        if (m_serialCombo->count() == 0) return;
        QString port = m_serialCombo->currentData().toString();
        auto& cfg = ConfigLoader::instance();
        SerialConfig sCfg = cfg.getSerialConfig();
        m_recv->startSerial(port, sCfg.baudrate, sCfg.protocol);
        m_serialToggle->setText("关闭串口");
        m_serialToggle->setStyleSheet(Styles::STYLE_BTN_ACTIVE());
        m_serialCombo->setEnabled(false);
    }
}

// 切换 UDP 接收状态
void ToolBar::toggleUdp() {
    if (!m_recv) return;

    if (m_recv->isUdpRunning()) {
        m_recv->stopUdp();
        m_udpToggle->setText("接收UDP");
        m_udpToggle->setStyleSheet(Styles::STYLE_BTN_IDLE());
        m_udpPortSpin->setEnabled(true);
    } else {
        auto& cfg = ConfigLoader::instance();
        UdpConfig uCfg = cfg.getUdpConfig();
        int port = m_udpPortSpin->value();
        m_recv->startUdp(uCfg.ip, port);
        m_udpToggle->setText("停止接收");
        m_udpToggle->setStyleSheet(Styles::STYLE_BTN_ACTIVE());
        m_udpPortSpin->setEnabled(false);
    }
}

// 串口异步断开时恢复 UI 为空闲状态
void ToolBar::onSerialStopped() {
    setSerialUiIdle();
}

// 将串口相关控件恢复为未连接状态
void ToolBar::setSerialUiIdle() {
    m_serialToggle->setText("打开串口");
    m_serialToggle->setStyleSheet(Styles::STYLE_BTN_IDLE());
    m_serialCombo->setEnabled(true);
    emit serialStopRequested();
}

